src/bitcoin-wallet.cpp

```diff
@@ -0,0 +1,121 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
+#include <chainparams.h>
+#include <chainparamsbase.h>
+#include <consensus/consensus.h>
+#include <logging.h>
+#include <util/system.h>
+#include <util/strencodings.h>
+#include <wallet/wallettool.h>
+
+#include <stdio.h>
+
+const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
+
+static void SetupWalletToolArgs()
+{
+    SetupChainParamsBaseOptions();
+
+    gArgs.AddArg("-?", "This help message", false, OptionsCategory::OPTIONS);
+    gArgs.AddArg("-datadir=<dir>", "Specify data directory", false, OptionsCategory::OPTIONS);
+    gArgs.AddArg("-wallet=<wallet-name>", "Specify wallet name", false, OptionsCategory::OPTIONS);
+    gArgs.AddArg("-debug=<category>", "Output debugging information (default: 0).", false, OptionsCategory::DEBUG_TEST);
+    gArgs.AddArg("-printtoconsole", "Send trace/debug info to console (default: 1 when no -debug is true, 0 otherwise.", false, OptionsCategory::DEBUG_TEST);
+
+    gArgs.AddArg("info", "Get wallet info", false, OptionsCategory::COMMANDS);
+    gArgs.AddArg("create", "Create new wallet file", false, OptionsCategory::COMMANDS);
+
+    // Hidden
+    gArgs.AddArg("-h", "", false, OptionsCategory::HIDDEN);
+    gArgs.AddArg("-help", "", false, OptionsCategory::HIDDEN);
+}
+
+static bool WalletAppInit(int argc, char* argv[])
+{
+    SetupWalletToolArgs();
+    std::string error_message;
+    if (!gArgs.ParseParameters(argc, argv, error_message)) {
+        fprintf(stderr, "Error parsing command line arguments: %s\n", error_message.c_str());
+        return false;
+    }
+    if (argc < 2 || HelpRequested(gArgs)) {
+        std::string usage = strprintf("%s bitcoin-wallet version", PACKAGE_NAME) + " " + FormatFullVersion() + "\n\n" +
+                                      "wallet-tool is an offline tool for creating and interacting with Bitcoin Core wallet files.\n" +
+                                      "By default wallet-tool will act on wallets in the default mainnet wallet directory in the datadir.\n" +
+                                      "To change the target wallet, use the -datadir, -wallet and -testnet/-regtest arguments.\n\n" +
+                                      "Usage:\n" +
+                                     "  bitcoin-wallet [options] <command>\n\n" +
+                                     gArgs.GetHelpMessage();
+
+        fprintf(stdout, "%s", usage.c_str());
+        return false;
+    }
+
+    // check for printtoconsole, allow -debug
+    g_logger->m_print_to_console = gArgs.GetBoolArg("-printtoconsole", gArgs.GetBoolArg("-debug", false));
+
+    if (!fs::is_directory(GetDataDir(false))) {
+        fprintf(stderr, "Error: Specified data directory \"%s\" does not exist.\n", gArgs.GetArg("-datadir", "").c_str());
+        return EXIT_FAILURE;
```

Comment:
This is not defined behavior, since EXIT_FAILURE has no defined type nor value, but you need a bool(false). c.f. https://en.cppreference.com/w/c/program/EXIT_status

Reply:
I don't understand this. There are plenty of other locations where `return EXIT_FAILURE` is used (and even a couple of PRs where `return false` has been changed to `return EXIT_FAILURE` - #9067 and #13349)

Reply:
We are not in `main` here, which has a return type `int`. The function signature here is `static bool WalletAppInit`, so the return type is `bool`. Commonly `EXIT_FAILURE` is `1`, so `bool(1) == true`, which off-by-one.

You can also test this trivially by providing a datadir that does not exist. It will print an error, but continue and print garbage.

Reply:
Duh. Thanks! I only looked at the github snippet. Will fix.